### PR TITLE
Plugins: Add AngularMeta and AngularMeta.DisableDeprecationUIFeatures

### PR DESCRIPTION
**Pull request by xnyo**
**What is this feature?**

- Adds AngularMeta to plugins and exposes it in frontendsettings. This groups all angular-related plugin settings/flags, and it replaces the previous `AngularDetected` property
- `AngularDetected` has been moved into `AngularMeta`
- Added `DisableDeprecationUIFeatures` to `AngularMeta`, which is a per-plugin config toggle to disable angular UI deprecation features. Currently it is only passed to the frontend, the functionality hasn't been implemented yet.
  - To set `DisableDeprecationUIFeatures` to `true` for a given plugin, add the following to the config file:
```ini
[plugin.grafana-polystat-panel]
disable_angular_deprecation_ui = true
```

**Why do we need this feature?**

Angular deprecation, refer to #68974 

**Who is this feature for?**

Grafana users with Angular plugins

**Which issue(s) does this PR fix?**:

Fixes #

**Special notes for your reviewer:**

Please check that:
- [x] It works as expected from a user's perspective.
- [x] If this is a pre-GA feature, it is behind a feature toggle.
- [x] The docs are updated, and if this is a [notable improvement](https://grafana.com/docs/writers-toolkit/writing-guide/contribute-release-notes/#how-to-determine-if-content-belongs-in-a-whats-new-document), it's added to our [What's New](https://grafana.com/docs/writers-toolkit/writing-guide/contribute-release-notes/) doc.

### CODE REVIEW

File: pkg/plugins/manager/loader/loader.go [hunk 1]
```diff
@@ -192,19 +192,22 @@ func (l *Loader) loadPlugins(ctx context.Context, src plugins.PluginSource, foun
...
 
 			// Do not initialize plugins if they're using Angular and Angular support is disabled
-			if p.AngularDetected && !l.cfg.AngularSupportEnabled {
+			if p.AngularMeta.AngularDetected && !l.cfg.AngularSupportEnabled {
 				l.log.Error("Refusing to initialize plugin because it's using Angular, which has been disabled", "pluginID", p.ID)
 				continue
 			}
 		}
+		if settings, ok := l.cfg.PluginSettings[p.ID]; ok {
+			p.AngularMeta.DisableDeprecationUIFeatures = settings["disable_angular_deprecation_ui"] == "true"
```
Comment (xnyo) [inline pkg/plugins/manager/loader/loader.go:209]: I'm not sure if this is the best way to do this (accessing l.cfg.PluginSettings map directly from the loader). By taking a quick look I couldn't find a better way but maybe I am missing something

File: pkg/plugins/models.go [hunk 2]
```diff
@@ -211,19 +211,20 @@ type PluginMetaDTO struct {
...
+	Name       string                 `json:"name"`
+	PluginMeta *PluginMetaDTO         `json:"meta"`
+	URL        string                 `json:"url,omitempty"`
+	IsDefault  bool                   `json:"isDefault"`
+	Access     string                 `json:"access,omitempty"`
+	Preload    bool                   `json:"preload"`
+	Module     string                 `json:"module,omitempty"`
+	JSONData   map[string]interface{} `json:"jsonData"`
+	ReadOnly   bool                   `json:"readOnly"`
+
```
Comment (academo) [inline pkg/plugins/models.go:226]: NIT: do you want this new line? (here and other places)

**Review by academo — APPROVED** [review]
LGTM. Maybe allow a BE dev to take a closer look to the GO code.

File: pkg/plugins/manager/loader/loader.go (hunk 1 above)
Comment (andresmgot) [reply pkg/plugins/manager/loader/loader.go:209]: I guess this is fine, I don't see an issue with it (only the naming).

**Review by andresmgot — COMMENTED** [review]
Code wise I only have a minor comment regarding the naming but feature wise does this make sense? A grafana administrator having to change the configuration to dismiss a warning per plugin seems a bit complicated. If it's more than just a warning (like other UI changes), why would an administrator choose to not display it (and only for a certain plugin)?

**Comment by xnyo** [comment]
@andresmgot thanks for the review :pray: . Answer to your question: some angular deprecation alerts will be dismissable (we still have to decide if per org, per user, per dashboard, etc) but in some cases, a Grafana administrator may acknowledge that they are affected by Angular deprecation and may want to silently disable all the angular deprecation UI elements for a particular plugin, so that users are not distracted by them.

See https://github.com/grafana/grafana/pull/70951 for a preview of what those (more invasive) elements may look like.

We have it in the config file so that they cannot do turn them off by accident (and complain that their dashboards stop working after the upgrade to a version of Grafana that doesn't support Angular anymore). We _try_ to make sure they know what they are doing.

It can also be useful for the support team, in case there are some false detections and they quickly want to dismiss those alerts for a particular instance that's affected, without waiting for us to update the remote detections in GCOM.

It effectively will act as a global kill-switch for all the angular deprecation UI features, if that makes sense.

File: packages/grafana-data/src/types/plugin.ts [hunk 3]
```diff
@@ -81,7 +81,12 @@ export interface PluginMeta<T extends KeyValue = {}> {
   signatureType?: PluginSignatureType;
   signatureOrg?: string;
   live?: boolean;
-  angularDetected?: boolean;
+  angularMeta?: PluginAngularMeta;
+}
+
+export interface PluginAngularMeta {
+  angularDetected: boolean;
+  disableDeprecationUIFeatures: boolean;
```
[the remark this answers is not in the document]
Comment (xnyo) [reply packages/grafana-data/src/types/plugin.ts:89]: Ah yes, agree. I will change it to `detected` and will think of a shorter name for the other one

**Comment by andresmgot** [comment]
> a Grafana administrator may acknowledge that they are affected by Angular deprecation and may want to silently disable all the angular deprecation UI elements for a particular plugin, so that users are not distracted by them.

I guess we are entering the UX space here but IMO we are putting on the patch before the wound :) Also, I think we should not add something that is so anoying that administrators need to be bothered to remove warnings. If we add that anyway, I would do it globally for all plugins so there is no need to go one by one (which I understand why you wouldd want to do that but it's more annoying).

I guess this is more feedback for @sympatheticmoose. Should we wait to see if anyone complains about the warnings before adding complexity to the project to bypass it?

**Comment by sympatheticmoose** [comment]
@andresmgot fair comments 👍 

By adding warnings within dashboards it becomes highly visible, but also disrupts the UX. It is a difficult balance given the urgency we wish to impose on users to migrate. I may be being overly-cautious on the requirement for it being on a per-plugin basis, as @xnyo states it was originally envisioned as a failsafe for false positives (including for private plugins where we have no easy way to correct). I worry that just disabling globally will then make it more likely that some dependencies will remain undiscovered. Maybe additional features in the EPIC help to mitigate this risk.

File: packages/grafana-data/src/types/plugin.ts (hunk 3 above)
[the remark this answers is not in the document]
Comment (andresmgot) [reply packages/grafana-data/src/types/plugin.ts:89]: maybe `disableWarnings` for the other one?

File: pkg/plugins/manager/loader/loader.go (hunk 1 above)
Comment (andresmgot) [reply pkg/plugins/manager/loader/loader.go:209]: here it would be `disable_angular_warnings`

**Review by andresmgot — APPROVED** [review]
IMO, adding complexity for fixing a problem we don't know if we will have is not worth it but I don't want to block this (since it's already implemented anyway :shrug:)

Code wise it looks good! I only have a suggestion for the name, feel free to disregard.

**Pull request closed without merging**